python/lsst/pipe/tasks/finalizeCharacterization.py

```diff
@@ -0,0 +1,693 @@
+
+        Parameters
+        ----------
+        input_schema : `lsst.afw.table.Schema`
+            Input schema.
+
+        Returns
+        -------
+        mapper : `lsst.afw.table.SchemaMapper`
+            Schema mapper
+        selection_schema : `lsst.afw.table.Schema`
+            Selection schema (with alias map)
+        """
+        mapper = afwTable.SchemaMapper(input_schema)
+        fields = input_schema.extract('*')
+        for field, item in fields.items():
+            mapper.addMapping(item.key)
+
+        mapper.editOutputSchema().addField(
+            'final_psf_candidate',
+            type='Flag',
+            doc=('set if the source was a candidate for PSF determination, '
+                 'as determined from FinalizeCharacterizationTask.'),
+        )
+        mapper.editOutputSchema().addField(
+            'final_psf_reserved',
+            type='Flag',
+            doc=('set if source was reserved from PSF determination by '
+                 'FinalizeCharacterizationTask.'),
+        )
+        mapper.editOutputSchema().addField(
+            'final_psf_used',
+            type='Flag',
+            doc=('set if source was used in the PSF determination by '
+                 'FinalizeCharacterizationTask.'),
+        )
+
+        selection_schema = mapper.getOutputSchema()
+        selection_schema.setAliasMap(input_schema.getAliasMap())
+
+        return mapper, selection_schema
+
+    def _concat_isolated_star_cats(self, band, isolated_star_cat_dict, isolated_star_source_dict):
+        """
+        Concatenate isolated star catalogs and make reserve selection.
+
+        Parameters
+        ----------
+        band : `str`
+            Band name.  Used to select reserved stars.
+        isolated_star_cat_dict : `dict`
+            Per-tract dict of isolated star catalog references.
+        isolated_star_source_dict : `dict`
+            Per-tract dict of isolated star source catalog references.
+
+        Returns
+        -------
+        isolated_table : `np.ndarray` (N,)
+            Table of isolated stars, with indexes to isolated sources.
+        isolated_source_table : `np.ndarray` (M,)
+            Table of isolated sources, with indexes to isolated stars.
+        """
+        isolated_tables = []
+        isolated_sources = []
+        merge_cat_counter = 0
+        merge_source_counter = 0
+
+        for tract in isolated_star_cat_dict:
+            df_cat = isolated_star_cat_dict[tract].get()
+            table_cat = df_cat.to_records()
+
+            df_source = isolated_star_source_dict[tract].get(
+                parameters={'columns': [self.config.id_column,
+                                        'obj_index']}
+            )
+            table_source = df_source.to_records()
+
+            # Cut isolated star table to those observed in this band, and adjust indices
+            use_band, = np.where(table_cat[f'nsource_{band}'] > 0)
```

Comment:
I use `where` because of my old idl days.  I'll see if I can retrain myself after decades.

And yes, fancy indexing is confusing which is why I tried to write a comprehensive test for it.  But it performs really well.